src/PubSubClient.cpp

```diff
@@ -214,18 +214,15 @@ bool PubSubClient::connect(const char* id, const char* user, const char* pass, c
 bool PubSubClient::connected() {
     if (!_client) return false;
 
-    bool rc = (bool)_client->connected();
-    if (!rc) {
-        if (_state == MQTT_CONNECTED) {
-            DEBUG_PSC_PRINTF("lost connection (client may have more details)\n");
-            _state = MQTT_CONNECTION_LOST;
-            _client->flush();
-            _client->stop();
-        }
-    } else {
-        return _state == MQTT_CONNECTED;
+    if (_client->connected()) {
+        return (_state == MQTT_CONNECTED);
+    } else if (_state == MQTT_CONNECTED) {
+        DEBUG_PSC_PRINTF("lost connection (client may have more details)\n");
+        _state = MQTT_CONNECTION_LOST;
+        _client->flush();
+        _client->stop();
```

Comment:
I was just checking what else is done when calling `disconnect()` and I think we're missing this one:
```c++
pingOutstanding = false;
```

Reply:
We can do this. But it won't change the current behavior. 
In `connected()` we set `_state = MQTT_CONNECTION_LOST`, so even if we reconnect the `_client` from outside we get a `false` here as the `_state` is not `MQTT_CONNECTED`. This will happen only, it we call `connect()`. And there 
https://github.com/hmueller01/pubsubclient3/blob/87c3e64c138b3533b6486cc90f43e8c407cb51b4/src/PubSubClient.cpp#L180
we set it `false` anyway ...